File: services/saga/orchestrator.ts
export interface SagaStep {
  name: string;
  execute: () => Promise<any>;
  compensate: () => Promise<void>;
}

export enum SagaStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  COMPENSATING = 'compensating',
  FAILED = 'failed',
}

export class Saga {
  private steps: SagaStep[] = [];
  private executedSteps: SagaStep[] = [];
  public status: SagaStatus = SagaStatus.PENDING;

  addStep(step: SagaStep): this {
    this.steps.push(step);
    return this;
  }

  async execute(): Promise<void> {
    this.status = SagaStatus.IN_PROGRESS;

    try {
      for (const step of this.steps) {
        await step.execute();
        this.executedSteps.push(step);
      }

      this.status = SagaStatus.COMPLETED;
    } catch (error) {
      console.error('Saga execution failed:', error);
      await this.compensate();
      this.status = SagaStatus.FAILED;
      throw error;
    }
  }

  private async compensate(): Promise<void> {
    this.status = SagaStatus.COMPENSATING;

    for (let i = this.executedSteps.length - 1; i >= 0; i--) {
      const step = this.executedSteps[i];

      try {
        await step.compensate();
      } catch (error) {
        console.error(`Compensation failed for step ${step.name}:`, error);
      }
    }
  }
}

export class SagaOrchestrator {
  async createOrderSaga(orderId: string, userId: string, items: any[]): Promise<void> {
    const saga = new Saga();

    saga
      .addStep({
        name: 'Reserve Inventory',
        execute: async () => {
          console.log('Reserving inventory for items:', items);
          return { reserved: true };
        },
        compensate: async () => {
          console.log('Releasing inventory reservation');
        },
      })
      .addStep({
        name: 'Process Payment',
        execute: async () => {
          console.log('Processing payment for order:', orderId);
          return { paymentId: 'pay_123' };
        },
        compensate: async () => {
          console.log('Refunding payment');
        },
      })
      .addStep({
        name: 'Update User Credits',
        execute: async () => {
          console.log('Updating user credits for user:', userId);
        },
        compensate: async () => {
          console.log('Reverting user credits');
        },
      })
      .addStep({
        name: 'Send Confirmation',
        execute: async () => {
          console.log('Sending order confirmation');
        },
        compensate: async () => {
          console.log('Sending cancellation notice');
        },
      });

    await saga.execute();
  }

  async createUserRegistrationSaga(
    email: string,
    username: string,
    password: string
  ): Promise<void> {
    const saga = new Saga();

    saga
      .addStep({
        name: 'Create User Account',
        execute: async () => {
          console.log('Creating user account:', username);
          return { userId: 'user_123' };
        },
        compensate: async () => {
          console.log('Deleting user account');
        },
      })
      .addStep({
        name: 'Send Welcome Email',
        execute: async () => {
          console.log('Sending welcome email to:', email);
        },
        compensate: async () => {
          console.log('No compensation needed for email');
        },
      })
      .addStep({
        name: 'Create Default Settings',
        execute: async () => {
          console.log('Creating default user settings');
        },
        compensate: async () => {
          console.log('Deleting user settings');
        },
      });

    await saga.execute();
  }
}

File: services/go/saga/orchestrator.go
package saga

import (
	"context"
	"fmt"
	"time"
)

type SagaStep struct {
	Name         string
	Action       func(ctx context.Context, data interface{}) error
	Compensation func(ctx context.Context, data interface{}) error
}

type SagaOrchestrator struct {
	steps      []SagaStep
	executed   []int
	sagaData   interface{}
}

func NewSagaOrchestrator(data interface{}) *SagaOrchestrator {
	return &SagaOrchestrator{
		steps:    make([]SagaStep, 0),
		executed: make([]int, 0),
		sagaData: data,
	}
}

func (s *SagaOrchestrator) AddStep(step SagaStep) {
	s.steps = append(s.steps, step)
}

func (s *SagaOrchestrator) Execute(ctx context.Context) error {
	for i, step := range s.steps {
		if err := step.Action(ctx, s.sagaData); err != nil {
			// Compensate all executed steps in reverse order
			if compensateErr := s.compensate(ctx); compensateErr != nil {
				return fmt.Errorf("saga failed at step %s and compensation failed: %w (original: %v)", 
					step.Name, compensateErr, err)
			}
			return fmt.Errorf("saga failed at step %s: %w", step.Name, err)
		}
		s.executed = append(s.executed, i)
	}
	return nil
}

func (s *SagaOrchestrator) compensate(ctx context.Context) error {
	// Execute compensations in reverse order
	for i := len(s.executed) - 1; i >= 0; i-- {
		stepIndex := s.executed[i]
		step := s.steps[stepIndex]
		
		if step.Compensation != nil {
			if err := step.Compensation(ctx, s.sagaData); err != nil {
				return fmt.Errorf("compensation failed for step %s: %w", step.Name, err)
			}
		}
	}
	return nil
}

// Example: Order Saga
type OrderSagaData struct {
	OrderID   string
	UserID    string
	ProductID string
	Amount    float64
	Reserved  bool
	Charged   bool
}

func CreateOrderSaga(data *OrderSagaData) *SagaOrchestrator {
	saga := NewSagaOrchestrator(data)

	// Step 1: Reserve inventory
	saga.AddStep(SagaStep{
		Name: "ReserveInventory",
		Action: func(ctx context.Context, d interface{}) error {
			orderData := d.(*OrderSagaData)
			fmt.Printf("Reserving inventory for product %s\n", orderData.ProductID)
			orderData.Reserved = true
			return nil
		},
		Compensation: func(ctx context.Context, d interface{}) error {
			orderData := d.(*OrderSagaData)
			fmt.Printf("Releasing inventory for product %s\n", orderData.ProductID)
			orderData.Reserved = false
			return nil
		},
	})

	// Step 2: Charge payment
	saga.AddStep(SagaStep{
		Name: "ChargePayment",
		Action: func(ctx context.Context, d interface{}) error {
			orderData := d.(*OrderSagaData)
			fmt.Printf("Charging $%.2f from user %s\n", orderData.Amount, orderData.UserID)
			orderData.Charged = true
			return nil
		},
		Compensation: func(ctx context.Context, d interface{}) error {
			orderData := d.(*OrderSagaData)
			fmt.Printf("Refunding $%.2f to user %s\n", orderData.Amount, orderData.UserID)
			orderData.Charged = false
			return nil
		},
	})

	// Step 3: Create order
	saga.AddStep(SagaStep{
		Name: "CreateOrder",
		Action: func(ctx context.Context, d interface{}) error {
			orderData := d.(*OrderSagaData)
			fmt.Printf("Creating order %s\n", orderData.OrderID)
			return nil
		},
		Compensation: func(ctx context.Context, d interface{}) error {
			orderData := d.(*OrderSagaData)
			fmt.Printf("Canceling order %s\n", orderData.OrderID)
			return nil
		},
	})

	return saga
}
